### ISSUE TITLE
request for named placeholder support

### ISSUE THREAD

**Issue by cobber** [sqlitebrowser/sqlitebrowser#2683]
Request: Support for Named Bind-Parameters
--------------------------

SQLite supports named bind-parameters, but there doesn't appear to be any way to use them interactively and they don't seem to be mentioned in sqlitebrowser's wiki.

When testing queries which are used in programs, often there will be *bind parameters* (a.k.a: *bind variables* or *placeholders*) which are bound to values at runtime. This allows a single query to be used for multiple scenarios without recompiling and also prevents [SQL injection](https://xkcd.com/327/) / data-based parsing problems.

e.g.:

`select id, name from employees where tax_class = `***:tax_class***`;`

or the more primitive

`select id, name from employees where tax_class = `***?***`;`

Additional info
---------------
#### Does this feature exist in another product or project? Please provide a link.
Toad for Oracle has this feature :-)
  - [Using Bind Variables](http://dev.toadfordataanalyst.com/webhelp/Content/Editor/Set_Bind_Variables.htm)
  - [video demonstration (from 0:45)](https://www.quest.com/video/how-to-create-and-utilize-parameters-in-queries-in-toad-data-point8127700/)

#### Do you have a screenshot? Please add screenshots to help explain your idea.
Provide an overview of any placeholders found in the query and allow the user to set/change their values before starting it. Ideally this would appear dynamically (in a side panel?) while editing the query, or at the very least as a pop-up just before the query is run.
e.g.:
`select id, name from employees where tax_class = `***:tax_class***` and dept = :department;`
| bind parameter | value |
|-|-|
| :tax_class | [ 123____ ] |
| :department | [ widgets ] |

#### Caching
It would also be great if sqlitebrowser would remember previously used `parameter -> value` pairs so that you don't have to start from scratch each time. (I have worked on projects which had queries with well over 10 parameters each)

All `parameter => value` pairs can be cached without regard to their respective queries.

Cached values should be kept permanently (i.e. they should survive restarts).

An option to clear all/some parameters from the cache would be nice-to-have.

**Comment by cobber** [comment]
### Additional info: `sqlite_parameters` table

I just discovered that a temporary table called `sqlite_parameters` can be used to provide parameter values. See [SQL Parameters in CLI](https://sqlite.org/cli.html#sql_parameters).

However, when I try to follow the steps provided, I just get errors:

### 1. manually create the parameters table:
```
create temp table sqlite_parameters(
  key text primary key,
  value any
) without rowid;

insert into sqlite_parameters values ( ':a', 1 ), ( ':b', 'foo' );

select :a as param_a, :b as param_b;
```
#### Error:
```
Execution finished with errors.
Result: object name reserved for internal use: sqlite_parameters
At line 1:
CREATE TEMP TABLE sqlite_parameters(
```

#### Note: this error is also produced by the sqlite3 CLI - so SQLite's documentation doesn't match its implementation here.

### 2. via SQLite's `.parameter` CLI command:
```
.parameter init
.parameter set :a 1
.parameter set :b foo
select :a as param_a, :b as param_b;
```
#### Error:
```
Execution finished with errors.
Result: near ".": syntax error
At line 1:
.
```

This, however, does work in the sqlite CLI.
```
sqlite> .parameter init
sqlite> .parameter set :a 1
sqlite> .parameter set :b foo
sqlite> select :a as param_a, :b as param_b;
param_a|param_b
1|foo
```

Cheers and thanks in advance!

**Comment by chrisjlocke** [comment]
You don't create the parameters table - that's what `.parameter init` does.

> Execution finished with errors.

The parameters table is only for SQLite CLI.  Support via the API is not supported.  I assume we talk to SQLite (in DB4S) via a library, which will use the API, so via the CLI.

> The temp.sqlite_parameters table only provides values for parameters in the command-line shell. The temp.sqlite_parameter table has no effect on queries that are run directly using the SQLite C-language API.

**Comment by cobber** [comment]
> You don't create the parameters table - that's what `.parameter init` does.

I understood "If there exists a TEMP table named `sqlite_parameters`" and "The `.parameter` command exists to simplify managing this table"
to mean "both are valid options" - obviously your interpretation is more accurate :-)

> The parameters table is only for SQLite CLI.

Doh!? And I thought I had (almost) found a workaround :-(
At least we cleared it up quickly.

Maybe DB4S could store a similar table somewhere (e.g. `~/.config/db4s/bindings.db`), allowing it to replace bind variables in queries with the cached values as needed.

**Comment by chrisjlocke** [comment]
As DB4S is cross-platform, a .config location wouldn't be feasible - but it might be something else to put in the project file (each database can have a project file with it) but cached parameters for one database wouldn't be available for another database.  Maybe you could have a database that lives with the application itself, but that wouldn't work so nicely on Windows (most programs live in c:\program files which most 'great unwashed' users don't have write access to. )

**Comment by cobber** [comment]
disclaimer: I don't have, like or want windows

**Comment by cobber** [comment]
`~/.config` was just one option based on my *nix bias ;-)

DB4S already stores user preferences *somewhere*, couldn't they be extended to hold a few key-value pairs?

**Comment by d4h0** [comment]
This would be a fantastic addition, and would transform `sqlitebrowser` into an IDE for writing SQLite queries. 

Right now, when we load an SQL file into `sqlitebrowser` (which is also used by our application), we have to replace ever query parameter with an actual value to work on a query, and then later we have to replace that value with a query parameter placeholder again (which we hopefully don't forget).

For me, this makes `sqlitebrowser` unfortunately  unusable for his task, which is a shame because `sqlitebrowser` would be amazing for writing SQL queries.

**Comment by Ciantic** [comment]
Old issue, but this is what I was looking for too.

The use-case is that you already have bunch of queries with placeholders like `?`

Now if I want to use those queries I have to modify the ? by hand instead of just pasting the values for placeholders.

It might seem like a minor thing, but if one has multiple queries it starts to add up.
